demo/llama/strategies.py

```diff
@@ -0,0 +1,241 @@
+import time
+from pathlib import Path
+from typing import Any, Callable, Dict, Generator, Optional, Union
+
+import torch
+import torch.distributed as dist
+from dataflux_core import user_agent
+from dataflux_pytorch.lightning import DatafluxLightningCheckpoint
+from dataflux_pytorch.lightning.gcs_filesystem import (GCSDistributedReader,
+                                                       GCSDistributedWriter)
+from google.cloud import storage
+from lightning.fabric.strategies import FSDPStrategy
+from lightning.fabric.utilities.types import _PATH, _Stateful
+from lightning.pytorch.strategies.fsdp import _METADATA_FILENAME
+from lightning.pytorch.trainer.states import TrainerFn
+from torch.distributed.checkpoint import async_save, load, save
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+from torch.nn import Module
+from torch.optim import Optimizer
+from typing_extensions import TypeGuard
+
+
+class DatafluxFSDPStrategy(FSDPStrategy):
+
+    def __init__(self,
+                 project_name,
+                 storage_client,
+                 use_async=False,
+                 **kwargs):
+        super().__init__(**kwargs)
+        self.ckpt_io = DatafluxLightningCheckpoint(project_name,
+                                                   storage_client)
+        self.storage_client = storage.Client(project=project_name)
+        user_agent.add_dataflux_user_agent(self.storage_client)
+
+        # Attributes used for async behavior.
+        self.use_async = use_async
+        self.checkpoint_group = None
+        self._checkpoint_future = None
+
+        if self.use_async:
+            default_ranks = list(range(dist.get_world_size()))
+            self.checkpoint_group = dist.new_group(
+                default_ranks, backend=self.process_group_backend)
+
+    def save_checkpoint(
```

Comment:
Add a link to save_checkpoint source code ? (Also I think this is coming from lightning fabric, any reason why we are using lightning fabric save_checkpoint instead of simple lightning checkpoint ?)

Reply:
lit-llama's training code does not use `lightning.Trainer` and `fabric` takes only custom strategies that inherit one of the classes defined in `lightning.fabric.strategies`. Another option I had was to re-write the training code to use lightning's `LightningModule`. This seemed simpler to do.